core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    private static Map<Integer, ViewVersion> indexVersions(List<ViewVersion> versionsToIndex) {
+      ImmutableMap.Builder<Integer, ViewVersion> builder = ImmutableMap.builder();
+      for (ViewVersion version : versionsToIndex) {
+        builder.put(version.versionId(), version);
+      }
+
+      return builder.build();
+    }
+
+    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
+      ImmutableMap.Builder<Integer, Schema> builder = ImmutableMap.builder();
+      for (Schema schema : schemasToIndex) {
+        builder.put(schema.schemaId(), schema);
+      }
+
+      return builder.build();
+    }
+
+    public ViewMetadata build() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
+      Preconditions.checkArgument(history.size() > 0, "Invalid view history: empty");
+      Preconditions.checkArgument(schemas.size() > 0, "Invalid schemas: empty");
+
+      Map<Integer, ViewVersion> versionsById = indexVersions(versions);
+      Preconditions.checkArgument(
+          versionsById.containsKey(currentVersionId),
+          "Cannot find current version %s in view versions: %s",
+          currentVersionId,
+          versionsById.keySet());
+
+      int currentSchemaId = versionsById.get(currentVersionId).schemaId();
+      Map<Integer, Schema> schemasById = indexSchemas(schemas);
+      Preconditions.checkArgument(
+          schemasById.containsKey(currentSchemaId),
+          "Cannot find current schema with id %s in schemas: %s",
+          currentSchemaId,
+          schemasById.keySet());
+
+      int versionHistorySizeToKeep =
+          PropertyUtil.propertyAsInt(
+              properties,
+              ViewProperties.VERSION_HISTORY_SIZE,
+              ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+
+      if (versionHistorySizeToKeep <= 0) {
```

Now that this is in the builder, we can throw an exception if it is invalid. Maybe convert this to a precondition?

---

makes sense, added a precondition and a test to ensure this fails properly